tests/test_aadt_mapping.py

```diff
@@ -18,19 +20,86 @@ def test_automapper():
     expect_not_found = ('Mul', 'Shftr', 'Shftl')
     for ir_name, ir_fc in IR.instructions.items():
         ir_mapper = arch_mapper.process_ir_instruction(ir_fc)
-        solution = ir_mapper.solve('z3')
-        if not solution.solved:
+        rewrite_rule = ir_mapper.solve('z3')
+        if rewrite_rule is None:
             assert ir_name in expect_not_found
             continue
         assert ir_name in expect_found
         #verify the mapping works
+        counter_example = rewrite_rule.verify()
+        assert counter_example is None
         ir_bv = ir_fc(family.PyFamily())
         for _ in range(num_test_vectors):
-            ir_vals = {path:BitVector.random(8) for path in solution.ir_bounded}
-            ir_inputs = solution.build_ir_input(ir_vals)
-            arch_inputs = solution.build_arch_input(ir_vals)
+            ir_vals = {path:BitVector.random(8) for path in rewrite_rule.ir_bounded}
+            ir_inputs = rewrite_rule.build_ir_input(ir_vals, family.PyFamily())
+            arch_inputs = rewrite_rule.build_arch_input(ir_vals, family.PyFamily())
             assert ir_bv()(**ir_inputs) == arch_bv()(**arch_inputs)
 
+def test_custom_rr():
+
+    #This is like a CoreIR Add
+    @family_closure
+    def coreir_add_fc(family):
+        Data = family.BitVector[16]
+        @family.assemble(locals(), globals())
+        class IR(Peak):
+            @name_outputs(out=Data)
+            def __call__(self, in0: Data, in1: Data):
+                return in0 + in1
+        return IR
+
+    #Simple PE that can only add or subtract
+    @family_closure
+    def PE_fc(family):
+        Data = family.BitVector[16]
+        class Inst(Enum):
+            add = 1
+            sub = 2
+
+        @family.assemble(locals(), globals())
+        class Arch(Peak):
+            def __call__(self, inst : Const(Inst), a: Data, b: Data) -> Data:
+                if inst == Inst.add:
+                    return a + b
+                else: #inst == Inst.sub
+                    return a - b
+        return Arch, Inst
+
+
+    Inst_bv = PE_fc(family.PyFamily())[1]
+    Inst_adt = AssembledADT[Inst_bv, Assembler, BitVector]
+
+    ir_fc = coreir_add_fc
+    arch_fc = lambda f: PE_fc(f)[0] #Only return peak class
+    output_binding = [(("out",), (0,))]
+
+    #Test correct rewrite rule
+    input_binding = [
+        (("in0",), ("a",)),
+        (("in1",), ("b",)),
+        (Inst_adt(Inst_bv.add)._value_, ("inst",)),
```

Comment:
Unfortunately this looks a bit gnarly. But this is currently how you get the Py BV value from the adt type.

Reply:
You could just assemble it by hand no?

Reply:
If you knew the exact bitvector (the case of Enum is easy) then yeah. Might be worth it to have a method on AADT that returns the underlying AbstractBitVector value

Reply:
I meant more just call `assembler.assemble(Inst_bv.add, BitVector)`  not sure if it's any more ergonomic.  Can't really add methods to AADTs without reserving field names so I would rather not.